### ISSUE TITLE
feature request: option to configure ember environment

### ISSUE THREAD

**Issue by jpaas** [thoughtbot/ember-cli-rails#287]
Currently it seems that the ember environment is determined based on the RAILS_ENV, or whether the `ember:test` rake task is being run but there doesn't seem to be any way to explicitly take control of it.

In our use case we have a suite of capybara acceptance tests for our ember app, but our ember app needs to be run in the test environment in order to lower some timeout values for testing. But I can't see any way of telling ember-cli-rails that I want my ember app run in that environment unless I'm missing something.

**Comment by seanpdoyle** [comment]
@jpaas https://github.com/thoughtbot/ember-cli-rails/blob/master/lib/ember_cli/helpers.rb#L25-L31 determines the current environment.

Does running with `Rails.env == "test"` not execute your Ember test suite in the `test` environment?

**Comment by jpaas** [comment]
@seanpdoyle I tried setting it up to check the environment, but every time I navigate to where I mount the ember app, it gives a 401 and redirects me to the root path. I suspect its some sort of devise issue. But I'm not sure why devise is getting involved.

**Comment by seanpdoyle** [comment]
The response has a status of `401` from a Rails route, or is your Ember app configured to error with a `401`?

If the `401` is coming from Rails, this doesn't sound like an `ember-cli-rails` issue.

**Comment by jpaas** [comment]
The 401 appears to be the response from EmberCli::EmberController#index

**Comment by seanpdoyle** [comment]
@jpaas ahh I see.

`EmberCli::EmberController` extends from `ApplicationController` (https://github.com/thoughtbot/ember-cli-rails/blob/master/app/controller/ember_cli/ember_controller.rb#L2).

Are you defining auth elsewhere?

**Comment by jpaas** [comment]
Ah ya and my ApplicationController has
```
  before_action :authenticate_user!
```

So I need to find a way to step around it when on the ember app path somehow.

**Comment by jpaas** [comment]
I just commented it out, so now its trying to serve the ember app, but the directory where its supposed to be in the tmp folder is empty :(

**Comment by seanpdoyle** [comment]
If you need to do something other than what `EmberCli::EmberController` is doing, you can route the ember request to a controller of your own:

https://github.com/thoughtbot/ember-cli-rails#configuring-the-ember-controller

**Comment by seanpdoyle** [comment]
@jpaas have you tried to configure the controller? 

Have you been able to make any progress?

**Comment by seanpdoyle** [comment]
@jpaas I'm closing this issue.

If this is still a problem, please say so and we can try again.

**Comment by jpaas** [comment]
Sorry for the delay. I was just able to get back to this. I am still totally stuck. I can't seem to get the ember app built.  I get a rack app error that it can't open my index.html from the app directory in my tmp folder. It creates a folder under tmp for my app, but the folder is always empty. It doesn't log anything to tell me that it wasn't able to build my ember app.

**Comment by seanpdoyle** [comment]
@jpaas that sounds awful.

What version of the gem are you using?

Have you tried the [latest for the gem](https://github.com/thoughtbot/ember-cli-rails/releases/latest) and [the addon](https://github.com/rondale-sc/ember-cli-rails-addon/releases/latest)?

**Comment by jpaas** [comment]
I'm using ember-cli-rails 0.5.6 and ember-cli-rails-addon 0.5.1

**Comment by jpaas** [comment]
Just tried moving up to ember-cli-rails-addon 0.5.2, but no difference.

I've created a custom controller to bypass authentication on my ApplicationController:
```
module Ember
  class HcpWebappController < ActionController::Base

    layout 'ember'

    def index
    end
  end
end
```

my routes.rb:
```
  if Rails.env.development? || Rails.env.test?
    mount_ember_app :hcp_webapp, to: '/ember/hcp_webapp', controller: 'ember/hcp_webapp'
  end
```

my views/ember/hcp_webapp/index.html.erb:
```
<script src="/assets/hcp_webapp/intl/polyfill/Intl.complete.js"></script>
<%= render_ember_app :hcp_webapp do |head| %>
    <% head.append do %>
        <title>hcp-webapp</title>
        <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
        <%= csrf_meta_tags %>
    <% end %>
<% end %>
```

my config/initializers/ember.rb:
```
EmberCLI.configure do |c|
  c.app :hcp_webapp,
        path: '../hcp-webapp'
end if defined?(EmberCLI)
```

**Comment by seanpdoyle** [comment]
On an unrelated note, you should include the polyfil `<script>` tag within the call to `render_ember_app`, since the current form would result in invalid HTML:

```erb
<%= render_ember_app :hcp_webapp do |head| %>
    <% head.append do %>
        <script src="/assets/hcp_webapp/intl/polyfill/Intl.complete.js"></script>
        <title>hcp-webapp</title>
        <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
        <%= csrf_meta_tags %>
    <% end %>
<% end %>
```

As far as the bug goes, everything looks like it should be working. I'll try looking into this further.

Would it be possible for you to create a sample repository that reproduces this behavior to make my debugging easier?

**Comment by jpaas** [comment]
@seanpdoyle I'll try and put together a pair of repos. One for the rails app and another for the separate ember-cli app.

**Comment by seanpdoyle** [comment]
@jpaas would you mind combining them? that would greatly simplify my process.

**Comment by jpaas** [comment]
@seanpdoyle sure, as long as that isn't part of the problem, the ember-cli repo being outside the rails repo.

**Comment by jpaas** [comment]
@seanpdoyle: well at least I am consistent when I mess up. Same result here https://github.com/jpaas/ember-rails-test

**Comment by seanpdoyle** [comment]
@jpaas I've pushed some fixes: 

https://github.com/jpaas/ember-rails-test/pull/1

**Comment by jpaas** [comment]
@seanpdoyle Thanks for all the help. I merged your PR and it worked. I then tried using #325 on my app and it now prints the following exception

```
ActionView::Template::Error (undefined method `encoding' for true:TrueClass):
  app/views/ember/hcp_webapp/index.html.erb:1:in `_app_views_ember_hcp_webapp_index_html_erb__1379877014170156989_70214761921720'
```

where line 1 is the `render_ember_app`

**Comment by seanpdoyle** [comment]
@jpaas very interesting.

It looks like you're silently failing to build the `index.html` file.

Would you mind pulling the latest from #325 and trying again?

**Comment by jpaas** [comment]
I'm not sure if this helps any more, but here's what I'm seeing:

```
Started GET "/ember/hcp_webapp" for ::1 at 2015-12-04 15:25:08 -0500
  ActiveRecord::SchemaMigration Load (0.4ms)  SELECT "schema_migrations".* FROM "schema_migrations"
Processing by Ember::HcpWebappController#index as HTML
  Parameters: {"ember_app"=>:hcp_webapp}
  Rendered ember/hcp_webapp/index.html.erb (16.9ms)
Completed 500 Internal Server Error in 26ms (ActiveRecord: 0.0ms)

ActionView::Template::Error (          EmberCLI failed to generate an `index.html` file.
):
    1: <%= render_ember_app :hcp_webapp do |head| %>
    2:     <% head.append do %>
    3:         <title>hcp-webapp</title>
    4:         <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
  app/views/ember/hcp_webapp/index.html.erb:1:in `_app_views_ember_hcp_webapp_index_html_erb___493391130970148090_70185104337560'
```

**Comment by seanpdoyle** [comment]
Does `Ember::HcpWebappController` extend from `ApplicationController`?

**Comment by jpaas** [comment]
No it extends from ActionController::Base. In my app the ApplicationController authenticates, and I didn't want the ember controller to authenticate. That's why I created a custom controller that extends ActionController::Base.

**Comment by seanpdoyle** [comment]
@jpaas would you be open to:

```rb
class Ember::HcpWebappController < ApplicationController
  skip_before_action :authenticate_user!
end
```

**Comment by jpaas** [comment]
@seanpdoyle I just tried extending ApplicationController with the skip_before_action but I get the same result

**Comment by seanpdoyle** [comment]
@jpaas is your `ApplicationController` an ancestor of `ActionController::Base`?

If so, I'm not sure how to proceed debugging this. Would you be willing to add me as a collaborator temporarily to help get to the bottom of this?

**Comment by jpaas** [comment]
The controllers look like this...

```ruby
class ApplicationController < ActionController::Base
  before_action :authenticate_user!
end
```

```ruby
module Ember
  class HcpWebappController < ApplicationController

    layout false
    skip_before_action :authenticate_user!

  end
end
```
The rails app and the ember app are in separate repos in Bitbucket. Do you have a bitbucket account?

**Comment by seanpdoyle** [comment]
I don't have a Bitbucket account.

What happens if you change to:

```rb
module Ember
  class HcpWebappController < ::ApplicationController # note the `::` prefix
    # ...
  end
end
```

**Comment by seanpdoyle** [comment]
I think what's happening is that the EmberCLI `before_action` is firing before the `authenticate_user!` before action, and that's what's causing issues

**Comment by jpaas** [comment]
@seanpdoyle I just tested both of those theories by adding :: and by commenting out the before_action entirely. The only message I ever get is "EmberCLI failed to generate an `index.html` file."

Once I had commented out the before_action I also tried reverting back to the user of the EmberCli controller instead of my custom one. But still it won't generate an index.html.

**Comment by seanpdoyle** [comment]
@jpaas I think I've got it!

Could you please try https://github.com/thoughtbot/ember-cli-rails/pull/327, changing your controller's `index` action to:

```rb
def index
  EmberCli["your-app"].build
  render layout: false
end
```

**Comment by jpaas** [comment]
Sorry @seanpdoyle  but that didn't work

**Comment by jpaas** [comment]
@seanpdoyle any other ideas or things I could do to help figure this out?

**Comment by seanpdoyle** [comment]
@jpaas I've joined BitBucket:

https://bitbucket.org/seanpdoyle/

Would you mind granting me access to the problem repository?

**Comment by jpaas** [comment]
@seanpdoyle sorry it took me a while, I had to free up a developer on our bitbucket plan to add you. I've given you access to 3 repositories. https://bitbucket.org/imdhealth/imdcloud/branch/solano is the branch of the rails app I've been using to muck about with ember-cli-rails. I've also given you access to the ember app repo which is included in the rails app as a git submodule. I've also given you access to an ember-addon repo that is a private addon used by the ember app. Phew. Sorry its so complicated.

**Comment by seanpdoyle** [comment]
```rb
# config/application.rb

# Force ember-cli-rails to run ember apps in a custom 'rails' environment in order to get the config we want

config.ember_cli_rails_mode = :rails
```

What is the behavior you're trying to achieve with this setting?

**Comment by jpaas** [comment]
Originally when I was trying to setup ember-cli-rails I tried to create a custom ember environment for when it was being run as part of ember-cli-rails. That way I could set the base_url based on wherever it got mounted. But I believe that environment was removed. Is that the problem? That setting is still hanging around and is messing things up?

**Comment by seanpdoyle** [comment]
@jpaas removing that configuration got the ball rolling. 

I don't have much time today to continue digging, but your project's setup is more complicated than the projects I've used the gem on, so it might take some time.

Your bug report uses `render_ember_app`, but your current source code uses the `include_` style helpers. Am I missing something?

**Comment by jpaas** [comment]
@seanpdoyle Yes I removed it and it got the ball rolling too. I had to run npm i && bower i (which incidentally is not mentioned in the README.md :) )

Thanks so much for having a look at this!

Where do you see these `include_` style helpers? I removed them in favour of `render_ember_app` as you suggested.

Now that its building I actually do want to run the ember app in a custom environment in order to set the base_url. Is `config.ember_cli_rails_mode` the way to do that?

**Comment by jpaas** [comment]
@seanpdoyle update: it is building the ember app now. I can see the symlink to the build in the tmp directory. But there are 2 problems:
1. On first load it always fails to load index.html (unable to open the file)
2. On second load it finds the index.html, but none of the assets are found. 404s:
```
GET http://localhost:3000/assets/vendor.css 
(index):64 GET http://localhost:3000/assets/intl/locales/en.js 
(index):63 GET http://localhost:3000/assets/intl/intl.min.js 
(index):17 GET http://localhost:3000/assets/hcp-webapp.css 
(index):65 GET http://localhost:3000/assets/vendor.js 
(index):2 GET http://localhost:3000/assets/hcp_webapp/intl/polyfill/Intl.complete.js 
(index):66 GET http://localhost:3000/assets/hcp-webapp.js 
```

**Comment by seanpdoyle** [comment]
The assets `404` are releated to your `ember-cli-build.js` manually disabling `fingerprint.enabled`, which the `addon` doesn't override if you've already set it:

https://github.com/rondale-sc/ember-cli-rails-addon/blob/a5d2cf050dce8d943f1eced990a535a8f6e7ac11/index.js#L48

**Comment by jpaas** [comment]
@seanpdoyle I tried commenting out the configuration of app.fingerprint entirely but it didn't give any different result

**Comment by seanpdoyle** [comment]
@jpaas would you mind trying out:

* https://github.com/thoughtbot/ember-cli-rails/tree/remove-sprockets for `ember-cli-rails` gem
* https://github.com/rondale-sc/ember-cli-rails-addon/tree/remove-sprockets for `ember-cli-rails-addon`

Each of those branches has up-to-date README's.

**Comment by seanpdoyle** [comment]
I'm hopeful that we can get this fix landed and stabilize things for you.

Having said that, we don't plan on supporting the `include_ember_*_tags` helpers in post-`1.0.0` versions of this gem.

It is likely they'll be extracted to a stable gem, but I highly recommend you migrate toward using the `render_ember_app` helper, which relies 100% on EmberCLI to generate the page's HTML.

**Comment by seanpdoyle** [comment]
@sevab would you mind trying out https://github.com/thoughtbot/ember-cli-rails/pull/347?

It removes the need for the consumer (you) to manually invoke `EmberCli.build`.

**Comment by jpaas** [comment]
I tried the two remove-sprockets branches, but got the following:
```
Started GET "/ember/hcp_webapp" for ::1 at 2015-12-10 14:38:48 -0500
  ActiveRecord::SchemaMigration Load (2.6ms)  SELECT "schema_migrations".* FROM "schema_migrations"
Processing by Ember::HcpWebappController#index as HTML
  Parameters: {"ember_app"=>:hcp_webapp}
  Rendered ember/hcp_webapp/index.html.erb (14.3ms)
Completed 500 Internal Server Error in 28ms (ActiveRecord: 0.0ms)

ActionView::Template::Error (          EmberCLI failed to generate an `index.html` file.
):
    1: <%= render_ember_app :hcp_webapp do |head| %>
    2:     <% head.append do %>
    3:         <title>hcp-webapp</title>
    4:         <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
  app/views/ember/hcp_webapp/index.html.erb:1:in `_app_views_ember_hcp_webapp_index_html_erb___4058448380498102536_70169595928960'
```

To try #347, I should use master along with 0.5.2 of addon?

**Comment by seanpdoyle** [comment]
@jpaas could you try `master` of both the gem and the addon?

**Comment by jpaas** [comment]
@seanpdoyle same result. It's definitely trying to build because there's a large pause once it hits the controller, but before the error.

**Comment by seanpdoyle** [comment]
@jpaas the second commit of https://github.com/seanpdoyle/ember-cli-rails-heroku-example/tree/remove-sprockets contains everything necessary to serve an Ember app. Could you try using that as a reference?

**Comment by jpaas** [comment]
@seanpdoyle I started following that checkin as a guide to eliminate any differences I had. It was only when I noticed that the remove-sprockets example uses the `include_ember_*` tags. So I tried switching back from `render_ember_app` and it started to work.

**Comment by seanpdoyle** [comment]
Yes, that example demonstrates how to use `include_ember_*_tags`.

I **STRONGLY** recommend you avoid that approach if you can. I'm curious to see where the `mount_ember_app` is failing for you.

**Comment by jpaas** [comment]
@seanpdoyle You've told me to avoid those tags a few times so I've been using `render_ember_app`. How we we diagnose what's going wrong with it?

**Comment by seanpdoyle** [comment]
@jpaas  [`0.6.0` has landed](https://github.com/thoughtbot/ember-cli-rails/releases/tag/v0.6.0) and removes all sprockets integration.

Also, `0.6.0` of the addon includes https://github.com/rondale-sc/ember-cli-rails-addon/commit/98f05c91ac71ee5963351ffe73e1238e8c29d59a, which should hopefully fix your `EmberCLI failed to generate an `index.html` file.` issue.

I'm closing this issue. 

Please try out the new version. I'm very, very hopeful your fix is in `0.6.0`.

If it's still a problem, can you open a new issue with up to date information?

**Comment by seanpdoyle** [comment]
@jpaas if you still need to configure your Ember environment, https://github.com/thoughtbot/ember-cli-rails/pull/366 should give you what you need.
